### PR TITLE
[Block Library - Post Terms]: Add dynamic variations of custom taxonomies

### LINKED ISSUE #32924: Post Terms Block: automatically add all public taxonomies as variations
Somewhat related to #32920 

If themes or plugins register custom public taxonomies to be used to display custom information on a site's frontend, I think it would make sense to offer that information via the Post Terms block, so one can use the block in the site editor to inject that information in their templates.

Plugins can register those custom variations on their end as well, but maybe offering them in the editor out of the box as long as the taxonomies as marked as public would help with all existing plugins out there.

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Resolves: https://github.com/WordPress/gutenberg/issues/32924

This PR defines `Post Terms` block variations dynamically and registers them server side, available since 5.8.

## Why?

Now we can display a list of all eligible taxonomies(built ins and custom ones).

## Testing Instructions
1. Register a custom taxonomy
2. In the editor insert the variation for it like `post {your taxonomy name}`.
3. Observe that if terms are set to the post you're testing with, they are displayed properly both in editor and front-end.

Alternatively just install a plugin with custom taxonomies already like `WooCommerce` and test with the product categories and product tags.

## Screenshots or screencast 

https://user-images.githubusercontent.com/16275880/160560147-ff32f816-481c-4abc-b08a-fc12cfee06c7.mov

### CODE REVIEW

**Comment by paaljoachim** [comment]
This is really helpful! Thank you @ntsekouras 

I am using a dev site where I am updating various tutorials that also use the custom post type "Tutorial".

Finding post terms. Or I could also have written Post categories or Post tags.
Screenshot 2022-03-30 at 15 37 13

### Post Tutorial  Categories:
Screenshot 2022-03-30 at 15 38 30

---

### Post Tutorial  Tags:
Screenshot 2022-03-30 at 15 38 38

-----

Adding some text. (backend)
Screenshot 2022-03-30 at 15 50 36

Frontend:
Screenshot 2022-03-30 at 15 49 22

Backend shows the layout just a little different compared to the frontend.

**Comment by Mamaduka** [comment]
This works great, Nik.

The `Post` prefix for all taxonomies feels a little odd to me. What do you think if we only prefix taxonomies registered for posts?

```php
in_array( 'post', $taxonomy->object_type, true )
```

It would be nice if we could limit variations based on the post-type context in the future.

**Comment by ntsekouras** [comment]
>The Post prefix for all taxonomies feels a little odd to me. What do you think if we only prefix taxonomies registered for posts?

For me the `post` here is the generic one but I don't have something better to suggest.. What I think though is that we need some kind of prefix for every variation and since taxonomies and post types are related many to many, this could get complicated really fast.

--cc @mtias @jasmussen for any thoughts about the names.

**Comment by mtias** [comment]
I'd remove the prefix and honor the label name. For posts, "tags" and "categories" would be easier to read.

**Comment by ntsekouras** [comment]
>I'd remove the prefix and honor the label name. For posts, "tags" and "categories" would be easier to read.

Without any prefix we will end up with duplicate block names. For example the `Categories` block which is a list of all categories and `Categories` from Post Terms that list the current `post's` categories.

**Comment by ntsekouras** [comment]
>I'd remove the prefix and honor the label name. For posts, "tags" and "categories" would be easier to read.

> > Without any prefix we will end up with duplicate block names. For example the Categories block which is a list of all categories and Categories from Post Terms that list the current post's categories.

Alternatively I guess we could update the other blocks to show `Categories lists` etc.. It would be good in general to add dynamic variations of those blocks as well for custom taxonomies.

**Comment by paaljoachim** [comment]
Should we get this into WordPress 6.0? 
As it would be helpful to be able to show CPT terms.

**Comment by paaljoachim** [comment]
@adamziel @gziolo @youknowriad 
Let's get some additional thoughts on this PR, and see if it is something that should be merged today for WordPress 6.0.

**Comment by jameskoster** [comment]
> Alternatively I guess we could update the other blocks to show Categories lists etc

Could another option would be to prohibit placing the Categories block inside a Query?

If we're going to include the post type in the block name we should probably update the placeholder content as well, IE this should read 'No product tags':

Screenshot 2022-04-08 at 13 21 05

That said, it would be interesting to eliminate all the prefixes. I've said this before but "Post Title" is kind of odd to see when working with a Product query, or a Page query for that matter. 

Like Titles, concepts such as categories and tags are universal, so I'm not sure they ever need prefixes? If we used generic block names they could behave agnostically. IE the Categories and Tags blocks could theoretically adapt based on the selected post type. Whereas with the current implementation, if you swapped the Query to display `product`s instead of `post`s you'd need to manually remove the Categories block and add the Product Categories one which seems unnecessarily convoluted.

**Comment by jasmussen** [comment]
> That said, it would be interesting to eliminate all the prefixes. I've said this before but "Post Title" is kind of odd to see when working with a Product query, or a Page query for that matter.

Yes, I just ran into something like this recently, such as inserting a Post Title on a Page template, or looking for "Archive Title" in all blocks when in fact the one I should've picked was "Query Title". There are probably some cases were we can't unify all these; indeed there's that curious case on the archive template where you want to show both the title of the archive, and individual item titles inside a query loop. Finally there may be a benefit of being able to style these individually in the Global Styles → Blocks panel, and give them different fonts, sizes, treatments, etc. 

So those are some of the thoughts to balance, but definitely agree that there's some simplification to have here.

**Comment by jameskoster** [comment]
> indeed there's that curious case on the archive template where you want to show both the title of the archive, and individual item titles inside a query loop

In that case I think 'Archive title' is okay because it doesn't lean in to any particular post type. It will behave consistently placed in either `category.html` or `taxonomy-product_cat-tshirts.html`.

For a while I've thought that a good rule of thumb could be to look at the underlying function name for guidance around these non-custom blocks. IE `the_content()` = Content. `the_archive_title()` = Archive Title.

**Comment by ntsekouras** [comment]
>Could another option would be to prohibit placing the Categories block inside a Query?

I think it can create confusion because we will end up with different blocks with the same name. Also outside a Query Loop block we will have these different blocks exposed in the inserter - think of a single post content or template.

**Comment by jameskoster** [comment]
Good point. In that case I like your idea above:

> Alternatively I guess we could update the other blocks to show Categories lists etc...

**Comment by ntsekouras** [comment]
By rebasing I have lost (temporally) the change of icon introduced here: https://github.com/WordPress/gutenberg/pull/40096. I'll probably need to add a filter for assigning different icons.

**Comment by ntsekouras** [comment]
I pushed an update that changes:
1. The title of `Categories` block to `Categories List`
2. The dynamic variations use taxonomy's label as title and the description has `Display the assigned ${taxonomy_label}` format.
3. Added a filter to assign `icons` for `categories and tags`, but the remaining ones use the block's default icon, which currently is the `categories` icon.

**Comment by Mamaduka** [comment]
Maybe for WP 6.1, we should propose new arguments for taxonomy registration:

* `labels.variation_name` - Use for variation registration.
* `show_in_variation` - Top level argument. Defaults to `show_in_rest`, but will allow developers more control.

File: packages/block-library/src/post-terms/index.php [hunk 1]
```diff
@@ -49,10 +49,48 @@ function render_block_core_post_terms( $attributes, $content, $block ) {
...
+	$built_ins         = array();
+	$custom_variations = array();
+
+	// Create and register the eligible taxonomies variations.
+	foreach ( $taxonomies as $taxonomy ) {
+		$variation = array(
+			'name'        => $taxonomy->name,
+			'title'       => $taxonomy->label,
+			/* translators: %s: taxonomy's label */
+			'description' => sprintf( __( 'Display the assigned %s' ), $taxonomy->label ),
```
Comment (mcsf) [inline packages/block-library/src/post-terms/index.php]: This isn't enough for good i18n. Quoting from an old [Calypso comment](https://github.com/Automattic/wp-calypso/pull/5346#issuecomment-218601883) of mine,

> the hardcoded line 'Would you like to create one?' […] doesn't work well with a lot of languages. A concrete example is gender — nouns like _page_ or _post_ may carry gender, and the pronoun one needs to reflect the gender. In Portuguese:
>
> - a post is _um artigo_, masculine; the hardcoded line would be: _Quer criar um?_
> - a page is _uma página_, feminine; the hardcoded line would be: _Quer criar uma?_
>
> German, French and indeed most Indo-European languages will have the same problem.

The same problem applies when considering taxonomies, which could differ in gender (e.g. [PT] _uma categoria_, _um produto_), or — in more inflected languages like Finnish — could be subjected to case inflection (e.g. [FI] nominative _kategoria_, partitive _kategoriaa_).

The proper way to create these strings would be to make them CPT labels... but that seems very heavy-handed for such a specific string. So, instead, I suggest crafting the strings in a way that frees them from inflection (e.g. something as unoriginal as _Display the assigned taxonomy: %s_).

Comment (ntsekouras) [reply packages/block-library/src/post-terms/index.php]: So many nuances with languages 😄 . I updated per your suggestion.

**Comment by mtias** [comment]
Should post types also do this automatically as Query variations? (i.e. a Films query block if I have a Films custom post type, etc)

**Comment by ntsekouras** [comment]
>Should post types also do this automatically as Query variations? (i.e. a Films query block if I have a Films custom post type, etc)

Yes, we can do this in a follow up.

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Comment by mburridge** [comment]
Added the Needs Dev Note label in case a dev note is needed for 6.1 release.

**Comment by ntsekouras** [comment]
>Added the Needs Dev Note label in case a dev note is needed for 6.1 release.

I'm not sure if one is needed.. This PR just creates some dynamic variations of Post Terms block and are available in the editor for users..

**Comment by mburridge** [comment]
Okay, thanks @ntsekouras. I've removed the label.

**Comment by sabernhardt** [comment]
Could we reconsider a prefix for the Post Terms block, perhaps something like "Terms:"?

I think that changing the original "Categories" block to "Categories List" is fine, but I'm concerned about giving its name to another block (at least in English). Selecting a block named "Categories" does something very different now, and that can surprise people who like to add categories lists in their sites. Confusion would be less common than I originally thought because the Post Terms block is not available in the Widget editors, yet it can happen in the Site and Post editors.

And maybe this discussion belongs in a new issue.

**Comment by paaljoachim** [comment]
I believe this issue would be interesting to know about:
[Block Library - Post Terms]: Add dynamic variations of feeds
https://github.com/WordPress/gutenberg/issues/65351

**Pull request merged**